ReadMe.txt
JFileSync, Version 3.0.0
Copyright (C) 2022-2023, Oliver Burgmaier
Copyright (C) 2002-2007, Jens Heidrich

Release Notes

1) Introduction
2) License and Usage Terms
3) Requirements
4) Installation and Application Start
5) Known Issues
6) Development Notes
7) Included Directories and Files
8) Contact and Further Information


1) Introduction

JFileSync is used to synchronize directories of (usually) two different file
systems. For instance, you have a laptop and a workstation and you want to
synchronize different directories on those machines in a platform-independent
way.

What you have to do, to use JFileSync for that purpose is (1) making the
directories of the laptop available for the workstation or vice versa (e.g.,
via NFS, Samba, or Windows File System Sharing) or using a JFS server to
access an external file system structure directly and (2) specifying an
appropriate configuration profile for JFileSync.

All functions of JFileSync can be controlled via the Graphical User Interface
(GUI). However, JFileSync provides full access to all features (apart from
plug-ins) via the command line interface. Call 'java -jar jfs.jar -help' to
get an overview of all possible command line options.


2) License and Usage Terms

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
St, Fifth Floor, Boston, MA, 02110-1301, USA

JFileSync uses libraries or parts of the following Open Source projects:
- Eclipse icons, which are provided under the Common Public License - v 1.0.

You can find a copy all licenses of JFileSync and the used libraries in the
'docs' sub directory of this distribution.


3) Requirements

- Java 5 Runtime Environment >= 1.5.0 (see 'http://java.sun.com')


4) Installation and Application Start

- Unzip the distribution file to a directory of your choice.
- For simplicity reasons a Windows batch file ('JFileSync.bat') and a Unix
  shell script ('JFileSync.sh') are available via the main distribution
  directory in order to start the application directly.
  ATTENTION: Because JFileSync is distributed as a Zip archive, Unix users
  will have to give executable rights manually before launching the
  application, e.g.: 'chmod a+rx JFileSync.sh'.
  If you have problems with the file encoding or line delimiters (when using
  an older version of the shell script) you will have to convert the script
  from DOS format to Unix format, e.g.: 'dos2unix JFileSync.sh'.
- If you do not want to use the start scripts you may also do the following:
  Enter the 'lib' sub directory of the distribution and call
  'java -jar jfs.jar' in order to start the Graphical User Interface or
  call 'java -jar jfs.jar -help' in order to get information about command
  line options. Depending in your Operating System a double click on
  'jfs.jar' will also launch the application.


5) Known Issues

- When using a JFS server, the default timeout is set to 5 seconds. When using
  a slow modem connection this may cause a lot of timeouts during
  synchronization. Therefore, you should increase the timeout (via
  the JFS server properties) to at least a minute (3.600.000 ms) before
  starting a synchronization via modem.
- Currently, the used protocol for file transfers via a JFS server is not
  interlocked several times so that the success of some operations on server
  side are unknown for the client. This will be changed in future releases in
  order to improve the robustness.
- The JSF server transmits Java objects as part of its protocol. This leads
  to an increased communication overhead (compared to, e.g., a plain text
  protocol).
- Before synchronization you may select a sub-set of the computed copy and
  delete statements. You may, for instance, de-select copying a directory, but
  select copying the files included in the directory. In this case, copying the
  files only will fail, because their directory is never created.


6) Development Notes

Required packages for JFileSync development (not included in the distribution):
- Java 5 SDK >= 1.5.0 (see 'http://java.sun.com')
- Apache Ant >= 1.6.2 (see 'http://ant.apache.org')

Used and therewith recommended development tools:
- Java 5 SDK 1.5.0
- Apache Ant 1.6.2
- Eclipse 3.1.0

Ant build targets:
- 'ant' creates a jar file named 'jfs.jar' to the 'lib' sub directory.
- 'ant clean-all' cleans the development directory.
- 'ant docs' creates the API documentation to the 'docs' sub directory.
- 'ant release' creates a JFS end-user release.
- 'ant snapshot' creates a JFS snapshot release.
- 'ant dev' creates a JFS developer release.
- 'ant reset' resets the testing directories after a synchronization test.
- 'ant test' starts the JFS application.
- 'ant test-native' starts the native language JFS application.
- 'ant test-server' tests the JFS server implementation.
- 'ant src-license' replaces the license text of all source code files.
- 'ant web-preview' creates a preview of the JFS web pages.
- 'ant clean-web' cleans the preview of the JFS web pages.


7) Included Directories and Files

The following structure describes all directories and files included in the
JFileSync end user releases containing only resources necessary to run the
application:
- docs: The program licenses, QA plan, and change history.
- lib: Necessary libraries in order to run the system.
- profiles: Sample user profiles.
- jfilesync-src.zip: The packed source code of the overall application.
- JFileSync.bat: Batch file to launch JFileSync from Windows.
- JFileSync.sh: Shell script to launch JFileSync from Unix/Linux.
- JFileSync.ico: Icon resource file containing the JFileSync logo.
- ReadMe.txt: This file.

The following structure describes all directories and files included in the
JFileSync developer releases (apart from hidden files) containing all
resources necessary to develop the application:
- classes: Temporary class file folder, generated by the Ant build file.
- docs: The program licenses, QA plan, and change history.
- lib: Necessary libraries in order to run the system.
- profiles: Sample user profiles.
- src: The source code of the overall application.
- test: A testing directory structure including test profiles.
- JFileSync.bat: Batch file to launch JFileSync from Windows.
- JFileSync.sh: Shell script to launch JFileSync from Unix/Linux.
- JFileSync.ico: Icon resource file containing the JFileSync logo.
- ReadMe.txt: This file.
- Version.txt: The version of the current release.


8) Contact and Further Information

You may submit improvement suggestions, patches, and other enhancements to
Jens Heidrich <jensheidrich@web.de>.

JFileSync is a SourceForge project and available via the following URL:
http://jfilesync.sourceforge.net/


docs/Changes.txt
JFileSync Change History
Version 3.0.0
- Added new context action to compare source and target file by configured
  compare application in settings.xml
- Added new context action to add filter by selected file/folder
- Added command line parameter "compare" to directly start compare after opening
  GUI 
- Added configurable application exit dialog. You can define to always close
  application without this dialog
- Added configurable time difference for date comparison to ignore gaps between
  summer and winter time.

Version 2.2 (2007/07/20)
- Added the following new features:
  - Added new dialog to specify local and external/server directory pairs.
  - System icons are shown for local files of the sync table.
  - German language pack for GUI.
  - You may specify file properties and the range for regular expression-based
    filters.
  - A properties windows may be invoked via the pop-up menu of the
    synchronization table if only one element is selected.
  - The root elements (directory pairs of your profile) are explicitly shown
    in the synchronization table (different icon and background color) and it
    is not possible to change actions of root elements any more.
  - The view on the synchronization table is updated on-the-fly during
    comparison and synchronization.
  - Added new option to control whether manual changes of actions in the
    synchronization table should be kept when switching between synchronization
    modes.
  - Manual changes of actions in the synchronization table are highlighted in
    blue color and may be reset (that is, set to the default action for the
    current synchronization mode) via the context menu.
  - Added debug mode control for the command line, removed test run mode
    completely.
  - Added a new profile attribute to control whether write protection of a
    file is replicated during synchronization.
  - Added new option to start and stop the JFS server as a (kind of) service.
- Improved the following existing features:
  - Increased the width of some dialogs.
  - Perform additional check for all GUI table row and column sizes.
  - Clear synchronization table selection when updating table.
- Fixed the following bugs:
  - The server base and pass phrase could not be set using the command line
    and the GUI.
  - Added missing translation string "Mode: " for JFSMainView.
  - Fixed equals method of JFSFilter. If a filter was specified, the
    synchronization table was updated every time the OK button of the profile
    manager was pressed (independent from performed profile changes).
  - Added dispose to close operation of info view.
  - Toogle details is on "Show details" whenever comparison or synchronization
    is started; that is, it is reset properly.
  - Prevent Null pointer exception when JFS command line server is started
    without proper "System.in".
  - Fixed need for X11 (graphics context) when started with "-nogui".
  - Show progress command line print only when there is something to show.
  - Clean configuration object at the beginning, if main method is used as
    part of service wrapper.
  - Fixed overwriting the history, if root element is not active.
- News behind the scene (developer-specific):
  - New class JFSPropertiesView for file properties.
  - JFSComparisonTable was renamed to JFSTableView. The renderer and mouse and
    action listener components were extracted to JFSTableRenderer and
    JFSTableListener, respectively.
  - Extracted view management from JFSTask to JFSTable. Only the current view
    is kept in memory. If a new view is selected, the current view is changed
    accordingly.
  - JFSRootElement extends JFSElement in order to store root directory pairs
    in the synchronization table. A root element also stores corresponding
    file producers (moved from JFSComparison) and the history.
  - The parent and all children are stored for each JFSElement so that the
    directory structure is available all the time. Furthermore, the root
    element to which each JFS element belongs is stored, too.
  - The synchronization actions can be computed for a single JFS element
    without having knowledge about the rest of the synchronization table.
    When a comparison is performed, the actions are computed on-the-fly.
  - Removed class JFSTask and moved its functionality to singletons JFSTable,
    JFSComparison, and JFSSynchronization. The comparison and synchronization
    are performed using the JFS table from now on and don't have to be called
    for each JFS directory pair separately, like it was before.
  - Corrected lower/upper case in all properties files.


Version 2.1 (2005/10/08)
- Added the following new features:
  - You may change the performed actions for elements of the main table via
    a popup menu. You may also activate and deactivate the table elements.
    The mechanism works for multiple interval selections.
  - Added new action in order to delete source file as well as target file
    (must be set manually for selected files).
  - JFS displays a warning in case of not stored profile settings when loading
    a new one.
  - Added three new views on the main action table in order to exclusively view
    files to copy, files to delete, and ambiguous files.
  - Added button to re-browse already added directory pairs in the profile
    manager.
- Improved the following existing features:
  - Simplified user interface before confirming synchronization start and
    viewing synchronization progress. You may show and hide synchronization
    details.
  - The histories are sorted according to the last synchronization date.
  - Unused history files are automatically deleted from the configuration
    directory.
  - The Unix shell script uses Unix line delimiters and is UTF-8 encoded.
  - The previous JFS session is only restored (in graphical mode) if a non

src/jfs/JFileSync.java
/*
 * JFileSync
 * Copyright (C) 2002-2007, Jens Heidrich
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA, 02110-1301, USA
 */

package jfs;

import java.io.File;
import java.io.PrintStream;
import java.net.URL;

import javax.swing.UIManager;

import jfs.conf.JFSConfig;
import jfs.conf.JFSConst;
import jfs.conf.JFSDirectoryPair;
import jfs.conf.JFSFilter;
import jfs.conf.JFSLog;
import jfs.conf.JFSSettings;
import jfs.conf.JFSText;
import jfs.gui.JFSMainView;
import jfs.server.JFSServerFactory;
import jfs.shell.JFSShell;
import jfs.sync.JFSTable;

/**
 * JFileSync is an application for synchronizing pairs of directories. This
 * class is the main class of the application. It processes all command line
 * options and starts (1) a JFS server, (2) the Java Swing-based graphical user
 * interface, or (3) a command line shell.
 * 
 * @see JFSServerFactory#startCmdLineServer()
 * @see jfs.gui.JFSMainView
 * @see JFSShell#startShell(boolean)
 * @author Jens Heidrich
 * @version $Id: JFileSync.java,v 1.40 2007/07/20 15:24:22 heidrich Exp $
 */
public class JFileSync {

	/**
	 * Determines the home directory, where the JAR file respectively the class
	 * files are located.
	 */
	public static String getHome() {
		try {
			URL packageUrl = ClassLoader.getSystemResource("jfs");
			File jfsLibDir = null;

			// JFS started from JAR:
			if (packageUrl.getProtocol().equals("jar")) {
				URL jarUrl = new URL(packageUrl.getFile());
				File jfsFile = new File(jarUrl.toURI());
				jfsLibDir = jfsFile.getParentFile().getParentFile();
			}

			// JFS started from classes directory:
			if (packageUrl.getProtocol().equals("file")) {
				File jfsFile = new File(packageUrl.toURI());
				jfsLibDir = jfsFile.getParentFile();
			}

			return jfsLibDir.getPath();
		} catch (Exception e) {
			return ".";
		}
	}

	/**
	 * Start of the application.
	 * 
	 * @param args
	 *            Command line arguments.
	 */
	public static void main(String[] args) {
		// Get settings for the first time in order to load stored
		// settings before doing any actions:
		JFSSettings s = JFSSettings.getInstance();

		// Get translation and configuration object:
		JFSText t = JFSText.getInstance();
		JFSConfig config = JFSConfig.getInstance();
		PrintStream p = JFSLog.getOut().getStream();

		// Clean config before starting (if main method is used as service):
		config.clean();

		boolean quiet = false;
		boolean launchServer = false;
		boolean startService = false;
		boolean stopService = false;
		boolean loadDefaultFile = true;
		boolean nogui = false;
		boolean compare = false;
		int i = 0;

		// Handle command line arguments:
		try {
			while (i < args.length) {
				if (args[i].equals("-debug")) {
					s.setDebug(true);
				} else if (args[i].equals("-config")) {
					i++;
					config.load(new File(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-sync")) {
					i++;
					config.setSyncMode(Byte.parseByte(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-view")) {
					i++;
					config.setView(Byte.parseByte(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-dir")) {
					i++;
					config.addDirectoryPair(new JFSDirectoryPair(args[i],
							args[i + 1]));
					i++;
					loadDefaultFile = false;
				} else if (args[i].equals("-compare")) {
					compare = true;
				} else if (args[i].equals("-granularity")) {
					i++;
					config.setGranularity(Integer.parseInt(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-buffersize")) {
					i++;
					config.setBufferSize(Integer.parseInt(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-overwriteuseractions")) {
					config.setKeepUserActions(false);
					loadDefaultFile = false;
				} else if (args[i].equals("-nohistory")) {
					config.setStoreHistory(false);
					loadDefaultFile = false;
				} else if (args[i].equals("-nowriteprotection")) {
					config.setCanWrite(false);
					loadDefaultFile = false;
				} else if (args[i].equals("-include")
						|| args[i].equals("-exclude")) {
					i++;
					JFSFilter filter = new JFSFilter(args[i]);
					if (args[i - 1].equals("-include")) {
						config.addInclude(filter);
					} else {
						config.addExclude(filter);
					}
					if (((i + 1) < args.length) && !args[i + 1].startsWith("-")) {
						i++;
						filter.setType(args[i]);
						if (((i + 1) < args.length)
								&& !args[i + 1].startsWith("-")) {
							i++;
							filter.setRange(args[i]);
						}
					}
					loadDefaultFile = false;
				} else if (args[i].equals("-server")) {
					launchServer = true;
					startService = false;
					stopService = false;
					if (((i + 1) < args.length) && !args[i + 1].startsWith("-")) {
						i++;
						config.setServerBase(args[i]);
					}
					loadDefaultFile = false;
				} else if (args[i].equals("-service")) {
					launchServer = false;
					if (((i + 1) < args.length) && !args[i + 1].startsWith("-")) {
						i++;
						if (args[i].equals("start")) {
							startService = true;
							stopService = false;
						} else if (args[i].equals("stop")) {
							startService = false;
							stopService = true;
						} else {
							printCmdLineHelpAndExit();
						}
						if (((i + 1) < args.length)
								&& !args[i + 1].startsWith("-")) {
							i++;
							config.setServerBase(args[i]);
						}
					} else {
						printCmdLineHelpAndExit();
					}
					loadDefaultFile = false;
				} else if (args[i].equals("-port")) {
					i++;
					config.setServerPort(Integer.parseInt(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-passphrase")) {
					i++;
					config.setServerPassPhrase(args[i]);
					loadDefaultFile = false;
				} else if (args[i].equals("-timeout")) {
					i++;
					config.setServerTimeout(Integer.parseInt(args[i]));
					loadDefaultFile = false;
				} else if (args[i].equals("-shutdown")) {
					config.setServerShutDown(true);
					loadDefaultFile = false;
				} else if (args[i].equals("-nogui")) {
					nogui = true;
				} else if (args[i].equals("-quiet")) {
					quiet = true;
				} else if (args[i].equals("-laf")) {
					if (((i + 1) >= args.length) || args[i + 1].startsWith("-")) {
						p.println(t.get("cmd.laf"));

						UIManager.LookAndFeelInfo[] laf = UIManager
								.getInstalledLookAndFeels();

						for (int j = 0; j < laf.length; j++) {
							p.println(j + 1 + ". " + laf[j].getName() + " <"
									+ laf[j].getClassName() + ">");
						}

						System.exit(0);
					} else {
						i++;
						s.setLaf(args[i]);
					}
				} else if (args[i].equals("-help") || args[i].equals("-?")) {
					printCmdLineHelpAndExit();
				} else if (args[i].equals("-version")) {
					p.println(t.get("general.appName") + " "
							+ JFSConst.getInstance().getString("jfs.version"));
					p.println(t.get("info.copyright"));
					p.println(t.get("info.author"));
					p.println(t.get("info.home") + " " + getHome());
					System.exit(0);
				} else if (args.length == 1) {
					config.load(new File(args[i]));
					loadDefaultFile = false;
				} else {
					printCmdLineHelpAndExit();
				}

				i++;
			}
		} catch (ArrayIndexOutOfBoundsException e) {
			printCmdLineHelpAndExit();
		} catch (NumberFormatException e) {
			JFSLog.getErr().getStream().println(t.get("error.numberFormat"));
			System.exit(0);
		}

		// Start server/service and exit afterwards if server should be
		// launched:
		if (launchServer) {
			JFSServerFactory.getInstance().startCmdLineServer();
		} else if (startService) {
			JFSServerFactory.getInstance().startService();
		} else if (stopService) {
			JFSServerFactory.getInstance().stopService();
		} else {
			// Initialize new synchronization table:
			JFSTable table = JFSTable.getInstance();
			config.attach(table);

			if (!nogui) {
				// Start GUI:
				p.println(t.get("cmd.startGui"));

				// Determine whether the last configuration when (stored when
				// the program was exited should be loaded at GUI startup:
				s.setNoGui(false);
				JFSMainView view = new JFSMainView(loadDefaultFile);
				if (compare)
				{
					view.actionPerformed("COMPARE");
				}
			} else {
				s.setNoGui(true);
				JFSShell.startShell(quiet);
			}
		}
	}

	/**
	 * Performs a busy wait.
	 * 
	 * @param duration
	 *            The number of milli-seconds to wait.
	 */
	public static final void busyWait(long duration) {
		long time = System.currentTimeMillis();
		long stop = time + duration;
		while (time <= stop) {
			time = System.currentTimeMillis();
		}
	}

	/**
	 * Prints the command line help file and exits.
	 */
	private static final void printCmdLineHelpAndExit() {
		JFSShell.printURL(JFSConst.getInstance().getResourceUrl(
				"jfs.help.topic.cmdLine"));
		System.exit(0);
	}
}